js/src/common/components/Tooltip.tsx

```diff
@@ -0,0 +1,275 @@
+
+  private oldText: string = '';
+  private oldVisibility: boolean | undefined;
+
+  private shouldRecreateTooltip: boolean = false;
+  private shouldChangeTooltipVisibility: boolean = false;
+
+  view(vnode: Mithril.Vnode<TooltipAttrs, this>) {
+    /**
+     * We know this will be a ChildArray and not a primitive as this
+     * vnode is a component, not a text or trusted HTML vnode.
+     */
+    const children = vnode.children as Mithril.ChildArray | undefined;
+
+    // We remove these to get the remaining attrs to pass to the DOM element
+    const { text, tooltipVisible, showOnFocus = true, position = 'top', ignoreTitleWarning = false, html = false, delay = 0, ...attrs } = this.attrs;
+
+    if ((this.attrs as any).title && !ignoreTitleWarning) {
+      console.warn(
+        '`title` attribute was passed to Tooltip component. Was this intentional? Tooltip content should be passed to the `text` attr instead.'
+      );
+    }
+
+    const realText = this.getRealText();
+
+    // We need to recreate the tooltip if the text has changed
+    if (realText !== this.oldText) {
+      this.oldText = realText;
+      this.shouldRecreateTooltip = true;
+    }
+
+    if (tooltipVisible !== this.oldVisibility) {
+      this.oldVisibility = this.attrs.tooltipVisible;
+      this.shouldChangeTooltipVisibility = true;
+    }
+
+    // We'll try our best to detect any issues created by devs before they cause any weird effects.
+    // Throwing an error will prevent the forum rendering, but will be better at alerting devs to
+    // an issue.
+
+    if (typeof children === 'undefined') {
+      throw new Error(
+        `Tooltip component was provided with no direct child DOM element. Tooltips must contain a single direct DOM node to attach to.`
+      );
+    }
+
+    if (children.length !== 1) {
+      throw new Error(
+        `Tooltip component was either passed more than one or no child node.\n\nPlease wrap multiple children in another element, such as a <div> or <span>.`
+      );
+    }
+
+    const firstChild = children[0];
+
+    if (typeof firstChild !== 'object' || Array.isArray(firstChild) || firstChild === null) {
+      throw new Error(
+        `Tooltip component was provided with no direct child DOM element. Tooltips must contain a single direct DOM node to attach to.`
+      );
+    }
+
+    if (typeof firstChild.tag === 'string' && ['#', '[', '<'].includes(firstChild.tag)) {
+      throw new Error(
+        `Tooltip component with provided with a vnode with tag "${firstChild.tag}". This is not a DOM element, so is not a valid child element. Please wrap this vnode in another element, such as a <div> or <span>.`
+      );
+    }
+
+    this.firstChild = firstChild;
+
+    return children;
+  }
+
+  oncreate(vnode: Mithril.VnodeDOM<TooltipAttrs, this>) {
+    super.oncreate(vnode);
+
+    const domNode = (this.firstChild as Mithril.VnodeDOM<any, any>).dom as HTMLElement;
+
+    if (!domNode.isSameNode(this.childDomNode)) {
```

Comment:
Are these operations costly?

Reply:
It's identical to calling `===`, but with better browser support.

https://dom.spec.whatwg.org/#dom-node-issamenode